### DISCUSSION TITLE
Script to help me preserve unread filter

### DISCUSSION THREAD

**Discussion by nostodnayr** [joelekstrom/fastmate#76] [Q&A] [answered]
I want my mailboxes to all default to using the unread filter. I've tried doing this by adding the `filter=unread` query parameter into the mailbox link in the sidebar (thanks you for explaining [how to wait for an element to load before modifying it](https://github.com/joelekstrom/fastmate/issues/66#issuecomment-1026831933)). However, it seems Fastmail's code is doing stuff to make that happen and the URL changes as a back-up measure.

I'm open to working around it by modifying Fastmate's code to add the query parameter. Though I'd prefer if I could do it with a script. The trouble is, I have no idea what I'm doing with this level of JavaScript. Do you happen to know a way to do this or if it's possible? Please don't trouble yourself too much, though.

**Comment by joelekstrom** [comment, marked as answer]
Hey! I gave it a shot, this seems to work. With a caveat, it will now reload an extra time every time mailbox changes, which might get annoying. But it should be a starting point at least!

```js
// Add listeners for when current url changes
history.pushState = ( f => function pushState(){
    var ret = f.apply(this, arguments);
    window.dispatchEvent(new Event('pushstate'));
    window.dispatchEvent(new Event('locationchange'));
    return ret;
})(history.pushState);

history.replaceState = ( f => function replaceState(){
    var ret = f.apply(this, arguments);
    window.dispatchEvent(new Event('replacestate'));
    window.dispatchEvent(new Event('locationchange'));
    return ret;
})(history.replaceState);

window.addEventListener('popstate',()=>{
    window.dispatchEvent(new Event('locationchange'))
});

function currentMailbox() {
    var pathComponents = window.location.pathname.split("/")
    if (pathComponents[1] != "mail") { return null; }
    return pathComponents[2]
}

// When current mailbox changes, update searchParams with unread filter
var lastMailbox = currentMailbox();
window.addEventListener('locationchange', function(){
    var mailbox = currentMailbox();
    if (mailbox != null && mailbox != lastMailbox) {
        lastMailbox = mailbox;
        var searchParams = new URLSearchParams(window.location.search);
        searchParams.set("filter", "unread");
        window.location.search = searchParams.toString();
    }
})
```

**Reply by nostodnayr** [reply]
😍🤩🙏

This is brilliant! I can live with the reload flash if it spares me the extra click or two.

Thank you so much! 💜

**Reply by nostodnayr** [reply]
I had a look at this and can understand how you're doing it, so this was also instructive. (That also let me adjust it to not work apply the filter to All Mail/Archive).

Thanks again for taking the time to help me with this.

**Reply by joelekstrom** [reply]
Great to hear 😊 and thanks for using Fastmate!
